File: src/aoc2016/jav/Day22.java
package aoc2016.jav;

import java.awt.Point;
import java.io.IOException;
import java.nio.file.*;
import java.util.*;
import java.util.regex.*;
import java.util.stream.Collectors;

class Day22 {
  private static class Node extends Point {
    final int size;
    final int used;
    final int avail;

    public Node(int x, int y, int size, int used, int avail) {
      super(x, y);
      this.size = size;
      this.used = used;
      this.avail = avail;
    }
  }

  private static List<Node> parse(List<String> input) {
    return input.stream().map(s -> {
      Matcher m = Pattern.compile("(\\d+)-y(\\d+)\\s+(\\d+)T\\s+(\\d+)T\\s+(\\d+)").matcher(s);
      if (m.find())
        return new Node(Integer.valueOf(m.group(1)),
          Integer.valueOf(m.group(2)),
          Integer.valueOf(m.group(3)),
          Integer.valueOf(m.group(4)),
          Integer.valueOf(m.group(5)));
      else return null;
    }).filter(Objects::nonNull).collect(Collectors.toList());
  }

  private static long partOne(List<Node> s) {
    return s.stream().filter(n -> n.used > 0).mapToLong(n -> s.stream().filter(other -> !n.equals(other) && other.avail >= n.used).count()).sum();
  }

  private static long partTwo(List<Node> s) {
    int x_size = s.stream().max(Comparator.comparing(n -> n.x)).get().x;
    int y_size = s.stream().max(Comparator.comparing(n -> n.y)).get().y;
    Node wStart = null, hole = null;
    Node[][] nodes = new Node[x_size + 1][y_size + 1];
    s.forEach(n -> nodes[n.x][n.y] = n);
    for (int x = 0; x < nodes.length; x++) {
      for (int y = 0; y < nodes[x].length; y++) {
        Node n = nodes[x][y];
        if (x == 0 && y == 0)
          System.out.print("S");
        else if (x == x_size && y == 0)
          System.out.print("G");
        else if (n.used == 0) {
          hole = n;
          System.out.print("_");
        } else if (n.size > 250) {
          if (wStart == null)
            wStart = nodes[x - 1][y];
          System.out.print("#");
        } else
          System.out.print(".");
      }
      System.out.println();
    }
    assert hole != null;
    assert wStart != null;
    int result = Math.abs(hole.x - wStart.x) + Math.abs(hole.y - wStart.y);
    result += Math.abs(wStart.x - x_size) + wStart.y;
    return result + 5 * (x_size - 1);
  }

  public static void main(String[] args) throws IOException {
    List<String> s = Files.readAllLines(Paths.get("./input/2016/Day22_input.txt"));
    System.out.println("Part One = " + partOne(parse(s)));
    System.out.println("Part One = " + partTwo(parse(s)));
  }
}

File: src/aoc2016/kot/Day22.kt
package aoc2016.kot

import java.nio.file.Files
import java.nio.file.Paths
import java.util.regex.Pattern

object Day22 {
    data class Node(val x: Int, val y: Int, val size: Int, val used: Int, val avail: Int)

    fun parse(input: List<String>): List<Node> {
        return input.map {
            val m = Pattern.compile("(\\d+)-y(\\d+)\\s+(\\d+)T\\s+(\\d+)T\\s+(\\d+)").matcher(it)
            if (m.find())
                Node(m.group(1).toInt(), m.group(2).toInt(), m.group(3).toInt(), m.group(4).toInt(), m.group(5).toInt())
            else null
        }.filterNotNull()
    }

    fun partOne(s: List<Node>): Int {
        return s.filter { n -> n.used > 0 }.map { n -> s.filter { other -> n != other && other.avail >= n.used }.count() }.sum()
    }

    fun partTwo(s: List<Node>): Int {
        val xSize: Int = s.maxBy { it.x }!!.x
        val wallEdge: Node = s.find { it.x == s.find { it.size > 250 }!!.x - 1 }!!
        val emptyDisk: Node = s.find { it.used == 0 }!!
        val nodes = s.groupBy { it.x }

        for ((key, value) in nodes) {
            println(value.map {
                when {
                    it.x == 0 && it.y == 0 -> 'S'
                    it.x == xSize && it.y == 0 -> 'G'
                    it.size > 100 -> '#'
                    else -> '.'
                }
            })
        }
        var result = Math.abs(emptyDisk.x - wallEdge.x) + Math.abs(emptyDisk.y - wallEdge.y)
        result += Math.abs(wallEdge.x - xSize) + wallEdge.y + 5 * (xSize - 1)
        return result
    }
}

fun main(args: Array<String>) {
    val s = Files.readAllLines(Paths.get("./input/2016/Day22_input.txt"))
    println("Part One = ${Day22.partOne(Day22.parse(s))}")
    println("Part One = ${Day22.partTwo(Day22.parse(s))}")
}
